src/config.ts

```diff
@@ -1,9 +1,27 @@
-import {Config, ConfigFn} from '../types/config'
+import {Config, ConfigFn, QueryAllElements, QueryElement} from '../types/config'
 import {prettyDOM} from './pretty-dom'
 
 type Callback<T> = () => T
+
+const queryElement: QueryElement = <T extends Element>(
+  element: T,
+  selector: string,
+) => element.querySelector(selector)
+const queryElementAll: QueryAllElements = <T extends Element>(
+  element: T,
+  selector: string,
+) => element.querySelectorAll(selector)
+
 interface InternalConfig extends Config {
   _disableExpensiveErrorDiagnostics: boolean
+  /**
+   * Returns the first element that is a descendant of node that matches selectors.
+   */
+  queryElement: QueryElement
```

Comment:
What's the use case for `queryElement`? The tests only show `queryAllElement` usage.

Reply:
That's a tricky one. `queryElement` is currently only used by the label helpers. Now in theory the input might be placed in a different shadow dom than the label (for instance, because the input would be wrapped). However, this would lead to a broken link between the label and the element (see https://github.com/whatwg/html/issues/3219) and the test will fail anyways. This discussion has gained some traction again and has led to https://github.com/WICG/webcomponents/issues/917, in which case this function would be needed. 

Theoretically though, if I understand the logic in https://github.com/testing-library/dom-testing-library/blob/main/src/queries/label-text.ts#L142 correct, if you add the queryElement selector then the error message will be nicer. If you want I can play around and see if I can create a test that expect the failure, and add the constraint of the shadow boundary to the error message. The notice would be weird because it only works if you do query into the shadow dom.

Reply:
I added the case that would need the `queryElement` here: https://github.com/testing-library/dom-testing-library/pull/1054/commits/eabc513519fe0a690450e0ef1b036f4707ad829d 

As said, from a custom-element perspective it's a rather sad case atm as its supposed to fail